fastestimator/trace/metric/bleu_score.py

```diff
@@ -0,0 +1,95 @@
+# limitations under the License.
+# ==============================================================================
+from typing import Iterable, Union
+
+import numpy as np
+from nltk.translate.bleu_score import corpus_bleu
+
+from fastestimator.trace.meta.per_ds import per_ds
+from fastestimator.trace.trace import Trace
+from fastestimator.util.data import Data
+from fastestimator.util.traceability_util import traceable
+
+
+def get_hypotheses(input_data):
+    return [str(i) for i in input_data if i != 0]
+
+
+def get_reference(input_data):
+    return [get_hypotheses(input_data)]
+
+
+@per_ds
+@traceable()
+class BleuScore(Trace):
+    """Calculate the Bleu score for a nlp task and report it back to the logger.
+
+    Args:
+        true_key: Name of the key that corresponds to ground truth in the batch dictionary.
+        pred_key: Name of the key that corresponds to predicted score in the batch dictionary.
+        mode: What mode(s) to execute this Trace in. For example, "train", "eval", "test", or "infer". To execute
+            regardless of mode, pass None. To execute in all modes except for a particular one, you can pass an argument
+            like "!infer" or "!train".
+        ds_id: What dataset id(s) to execute this Trace in. To execute regardless of ds_id, pass None. To execute in all
+            ds_ids except for a particular one, you can pass an argument like "!ds1".
+        output_names: Name of the keys to store back to the state.
+        per_ds: Whether to automatically compute this metric individually for every ds_id it runs on, in addition to
+            computing an aggregate across all ds_ids on which it runs. This is automatically False if `output_name`
+            contains a "|" character.
+
+    """
+    def __init__(self,
+                 true_key: str,
+                 pred_key: str,
+                 mode: Union[None, str, Iterable[str]] = ("eval", "test"),
+                 ds_id: Union[None, str, Iterable[str]] = None,
+                 output_names: Iterable[str] = ('bleu_score_1', 'bleu_score_2', 'bleu_score_3', 'bleu_score_4'),
+                 per_ds: bool = True) -> None:
+        super().__init__(inputs=(true_key, pred_key), outputs=output_names, mode=mode, ds_id=ds_id)
+        assert len(self.outputs) == 4, 'BleuScore trace adds 3 fields bleu_score_1 bleu_score_2 bleu_score_3 bleu_score_4 to state dict'
+        self.list_of_references = []
+        self.hypotheses = []
+        self.per_ds = per_ds
+
+    @property
+    def true_key(self) -> str:
+        return self.inputs[0]
+
+    @property
+    def pred_key(self) -> str:
+        return self.inputs[1]
+
+    def on_epoch_begin(self, data: Data) -> None:
+        self.list_of_references = []
+        self.hypotheses = []
+
+    def on_batch_end(self, data: Data) -> None:
+        y_pred, y_true = np.argmax(data[self.pred_key].numpy(), axis=-1), data[self.true_key].numpy()
```

Comment:
should we have more careful handling here of true_key and pred_key in case they don't match these assumptions (as in Accuracy trace for example). Or is NLP guaranteed to be in this format?

Reply:
Currently we follow the same approach as accuracy with true_key and pred_key as a property values?

Reply:
sorry, should have been more explicit. I mean the shape assumption on true_key and pred_key (lines 74-83 in the accuracy trace). In Bleu the assumption being enforced is that the model is outputting a prob array that needs to be argmaxed, but what if the model already argmaxed? or what if the true_key array was a one-hot vector that hadn't been argmaxed yet...